File: openrndr-adopted-artifex/src/main/java/io/lacuna/artifex/utils/regions/Clip.java
package io.lacuna.artifex.utils.regions;

import io.lacuna.artifex.*;
import io.lacuna.artifex.Ring2.Result;
import io.lacuna.artifex.utils.Combinatorics;
import io.lacuna.bifurcan.*;

import java.util.Comparator;
import java.util.Objects;
import java.util.function.BiFunction;
import java.util.function.BinaryOperator;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static java.lang.Math.abs;

public class Clip {

  public static final BinaryOperator<Arc> SHORTEST_ARC = (x, y) -> x.length() < y.length() ? x : y;

  // for debug purposes

  private static final ISet<Vec2> VERTICES = new LinearSet<>();

  private static void describe(String prefix, IList<Vec2>... arcs) {
    for (IList<Vec2> arc : arcs) {
      arc.forEach(VERTICES::add);
      System.out.print(prefix + " ");
      arc.forEach(v -> System.out.print(VERTICES.indexOf(v) + " "));

      System.out.println();
    }
  }

  /**
   * an Arc is a list of curves, either describing a loop or an edge between two points of intersection
   */
  private final static class Arc extends LinearList<Curve2> {

    private double
      length = Double.NaN,
      area = Double.NaN;

    double length() {
      if (Double.isNaN(length)) {
        length = stream().mapToDouble(c -> c.end().sub(c.start()).length()).sum();
      }
      return length;
    }

    Vec2 head() {
      return first().start();
    }

    Vec2 tail() {
      return last().end();
    }

    Vec2 position(double t) {
      double length = length(),
        offset = 0,
        threshold = length * t;

      for (Curve2 c : this) {
        double l = c.end().sub(c.start()).length();
        Interval i = new Interval(offset, offset + l);
        if (i.contains(threshold)) {
          return c.position(i.normalize(threshold));
        }
        offset = i.hi;
      }

      throw new IllegalStateException();
    }

    Arc reverse() {
      Arc result = new Arc();
      forEach(c -> result.addFirst(c.reverse()));
      return result;
    }

    IList<Vec2> vertices() {
      IList<Vec2> result = new LinearList<Vec2>().addLast(head());
      forEach(c -> result.addLast(c.end()));
      return result;
    }

    double signedArea() {
      if (Double.isNaN(area)) {
        area = stream().mapToDouble(Curve2::signedArea).sum();
      }
      return area;
    }

    @Override
    public int hashCode() {
      return System.identityHashCode(this);
    }

    @Override
    public boolean equals(Object obj) {
      return this == obj;
    }
  }

  private static double area(IList<Arc> arcs) {
    return abs(arcs.stream().mapToDouble(Arc::signedArea).sum());
  }

  private static double length(IList<Arc> arcs) {
    return abs(arcs.stream().mapToDouble(Arc::length).sum());
  }

  private static Ring2 ring(IList<Arc> arcs) {
    IList<Curve2> acc = new LinearList<>();
    arcs.forEach(arc -> arc.forEach(acc::addLast));
    return new Ring2(acc);
  }

  private static Arc arc(Curve2... cs) {
    Arc result = new Arc();
    for (Curve2 c : cs) {
      result.addLast(c);
    }
    return result;
  }

  private static <U, V> IList<V> edges(IList<U> vertices, BiFunction<U, U, V> edge) {
    IList<V> result = new LinearList<>();
    for (int i = 0; i < vertices.size() - 1; i++) {
      result.addLast(edge.apply(vertices.nth(i), vertices.nth(i + 1)));
    }
    return result;
  }

  ///

  private enum Operation {
    UNION,
    INTERSECTION,
    DIFFERENCE
  }

  private enum Type {
    OUTSIDE,
    INSIDE,
    SAME_EDGE,
    DIFF_EDGE
  }

  private static boolean isTop(Curve2 c) {
    if (c == null) {
      return false;
    }

    double delta = c.end().x - c.start().x;
    if (delta == 0) {
      return c.end().y > c.start().y;
    }
    return delta < 0;
  }

  private static Type classify(Region2 region, Arc arc) {
    Result result = region.test(arc.position(0.5));
    if (!result.inside) {
      return Type.OUTSIDE;
    } else if (result.curve == null) {
      return Type.INSIDE;
    } else {
      return isTop(arc.first()) == isTop(result.curve) ? Type.SAME_EDGE : Type.DIFF_EDGE;
    }
  }

  /**
   * Cuts the rings of a region at the specified vertices, yielding a list of arcs that will serve as the edges of our
   * graph.
   */
  private static IList<Arc> partition(Region2 region, ISet<Vec2> vertices) {
    IList<Arc> result = new LinearList<>();

    for (Ring2 r : region.rings) {
      Curve2[] cs = r.curves;
      int offset = 0;
      for (; offset < cs.length; offset++) {
        if (vertices.contains(cs[offset].start())) {
          break;
        }
      }

      if (offset == cs.length) {
        result.addLast(arc(cs));
      } else {
        Arc acc = new Arc();
        for (int i = offset; i < cs.length; i++) {
          Curve2 c = cs[i];
          if (vertices.contains(c.start())) {
            if (acc.size() > 0) {
              result.addLast(acc);
            }
            acc = arc(c);
          } else {
            acc.addLast(c);
          }
        }

        for (int i = 0; i < offset; i++) {
          acc.addLast(cs[i]);
        }

        if (acc.size() > 0) {
          result.addLast(acc);
        }
      }
    }

    return result;
  }

  public static IList<IList<Arc>> greedyPairing(IGraph<Vec2, Arc> graph, IList<Vec2> out, ISet<Vec2> in) {
    IList<IList<Arc>> result = new LinearList<>();

    ISet<Vec2> remaining = LinearSet.from(in);
    for (Vec2 v : out) {
      IList<Vec2> path = Graphs.shortestPath(graph, v, remaining::contains, e -> e.value().length()).orElse(null);
      if (path == null) {
        return null;
      }

      remaining.remove(path.last());
      result.addLast(edges(path, graph::edge));
    }

    return result;
  }

  public static IList<IList<Arc>> repairGraph(IGraph<Vec2, ISet<Arc>> graph, Iterable<Arc> unused) {

    // create a graph of all the unused arcs
    IGraph<Vec2, Arc> search = new DirectedGraph<Vec2, Arc>().linear();
    for (Arc arc : unused) {
      search.link(arc.head(), arc.tail(), arc, SHORTEST_ARC);
    }

    // add in the existing arcs as reversed edges, so we can potentially retract them
    for (IEdge<Vec2, ISet<Arc>> e : graph.edges()) {
      Arc arc = e.value().stream().min(Comparator.comparingDouble(Arc::length)).get();
      search.link(arc.tail(), arc.head(), arc, SHORTEST_ARC);
    }

    //search.vertices().forEach(v -> System.out.println(VERTICES.indexOf(v) + " " + search.out(v).stream().map(VERTICES::indexOf).collect(Lists.linearCollector())));

    ISet<Vec2>
      out = graph.vertices().stream().filter(v -> graph.out(v).size() == 0).collect(Sets.linearCollector()),
      in = graph.vertices().stream().filter(v -> graph.in(v).size() == 0).collect(Sets.linearCollector()),
      currOut = out.clone(),
      currIn = in.clone();

    //describe("out", out.elements());
    //describe("in", in.elements());

    // attempt to greedily pair our srcs and dsts
    IList<IList<Arc>> result = new LinearList<>();
    while (currIn.size() > 0 && currOut.size() > 0) {
      IList<Vec2> path = Graphs.shortestPath(search, currOut, in::contains, e -> e.value().length()).orElse(null);

      // if our search found a vertex that was previously claimed, we need something better than a greedy search
      if (path == null || !currIn.contains(path.last())) {
        break;

      } else {
        currOut.remove(path.first());
        currIn.remove(path.last());
        result.addLast(edges(path, search::edge));
      }
    }

    if (currIn.size() == 0 || currOut.size() == 0) {
      return result;
    }

    // do greedy pairings with every possible vertex ordering, and choose the one that results in the shortest aggregate
    // paths
    return Combinatorics.permutations(out.elements())
      .stream()
      .map(vs -> greedyPairing(search, vs, in))
      .filter(Objects::nonNull)
      .min(Comparator.comparingDouble(paths -> paths.stream().mapToDouble(Clip::length).sum()))
      .get();
  }

  public static Region2 operation(Region2 ra, Region2 rb, Operation operation, Predicate<Type> aPredicate, Predicate<Type> bPredicate) {

    Split.Result split = Split.split(ra, rb);
    Region2 a = split.a;
    Region2 b = split.b;

    // Partition rings into arcs separated at intersection points
    IList<Arc>
      pa = partition(a, split.splits),
      pb = partition(b, split.splits);

    if (operation == Operation.DIFFERENCE) {
      pb = pb.stream().map(Arc::reverse).collect(Lists.linearCollector());
    }

    // Filter out arcs which are to be ignored, per our operation
    ISet<Arc> arcs = new LinearSet<>();
    pa.stream()
      .filter(arc -> aPredicate.test(classify(b, arc)))
      .forEach(arcs::add);
    pb.stream()
      .filter(arc -> bPredicate.test(classify(a, arc)))
      .forEach(arcs::add);

    //describe("split", split.splits.elements());
    //describe("arcs", arcs.elements().stream().map(Arc::vertices).toArray(IList[]::new));
    //VERTICES.forEach(v -> System.out.println(VERTICES.indexOf(v) + " " + v));

    IList<Ring2> result = new LinearList<>();
    ISet<Arc> consumed = new LinearSet<>();

    // First we're going to extract a cycle, and on the second go-around we'll try to "repair" the remaining edges, and
    // extract any additional cycles we create in the process.
    for (int i = 0; i < 2; i++) {

      // Construct a graph where the edges are the set of all arcs connecting the vertices
      IGraph<Vec2, ISet<Arc>> graph = new DirectedGraph<Vec2, ISet<Arc>>().linear();
      arcs.forEach(arc -> graph.link(arc.head(), arc.tail(), LinearSet.of(arc), ISet::union));

      //graph.vertices().forEach(v -> System.out.println(VERTICES.indexOf(v) + " " + graph.out(v).stream().map(VERTICES::indexOf).collect(Lists.linearCollector())));

      if (i == 1) {
        for (IList<Arc> path : repairGraph(graph, LinearSet.from(pa.concat(pb)).difference(arcs).difference(consumed))) {
          for (Arc arc : path) {
            // if the graph currently contains the arc, remove it
            if (arcs.contains(arc)) {
              //describe("remove", arc.vertices());
              graph.unlink(arc.head(), arc.tail());
              arcs.remove(arc);

              // if the graph doesn't contain the arc, add it
            } else {
              //describe("add", arc.vertices());
              graph.link(arc.head(), arc.tail(), LinearSet.of(arc));
              arcs.add(arc);
            }
          }
        }
      }

      // find every cycle in the graph, and then expand those cycles into every possible arc combination, yielding a bunch
      // of rings ordered from largest to smallest
      IList<IList<Arc>> cycles = Graphs.cycles(graph)
        .stream()
        .map(cycle -> edges(cycle, (x, y) -> graph.edge(x, y).elements()))
        .map(Combinatorics::combinations)
        .flatMap(IList::stream)
        .sorted(Comparator.comparingDouble(Clip::area).reversed())
        .collect(Lists.linearCollector());

      // extract as many cycles as possible without using the same arc twice
      for (IList<Arc> cycle : cycles) {
        //describe("cycle", cycle.stream().map(Arc::vertices).toArray(IList[]::new));
        if (cycle.stream().anyMatch(consumed::contains)) {
          continue;
        }

        cycle.forEach(consumed::add);
        result.addLast(ring(cycle));
      }

      arcs = arcs.difference(consumed);

      if (arcs.size() == 0) {
        break;
      }
    }

    return new Region2(result);
  }

  ///

  public static Region2 union(Region2 a, Region2 b) {
    return operation(a, b,
      Operation.UNION,
      t -> t == Type.OUTSIDE || t == Type.SAME_EDGE,
      t -> t == Type.OUTSIDE);
  }

  public static Region2 intersection(Region2 a, Region2 b) {
    return operation(a, b,
      Operation.INTERSECTION,
      t -> t == Type.INSIDE || t == Type.SAME_EDGE,
      t -> t == Type.INSIDE);
  }

  public static Region2 difference(Region2 a, Region2 b) {
    return operation(a, b,
      Operation.DIFFERENCE,
      t -> t == Type.OUTSIDE || t == Type.DIFF_EDGE,
      t -> t == Type.INSIDE);
  }

}

File: openrndr-kartifex/src/commonMain/kotlin/utils/regions/Clip.kt
package org.openrndr.kartifex.utils.regions

import org.openrndr.kartifex.Curve2
import org.openrndr.kartifex.Region2
import org.openrndr.kartifex.Ring2
import org.openrndr.kartifex.Vec2
import org.openrndr.kartifex.utils.Combinatorics
import org.openrndr.kartifex.utils.graphs.DirectedGraph
import org.openrndr.kartifex.utils.graphs.Graphs
import kotlin.math.E
import kotlin.math.abs

object Clip {
    // The approach used here is described at https://ideolalia.com/2018/08/28/artifex.html.  The "simplest" approach would
    // be to represent the unused segments as a multi-graph (since there can be multiple segments connecting any pair of vertices),
    // but the graph data structure used here is *not* a multi-graph, so instead we model it as a graph which only includes
    // the shortest edge between the vertices, and we just iterate over it multiple times.  Empirically 2-3 times should
    // always suffice, but we give ourselves a bit of breathing room because mostly we just want to preclude an infinite loop.
    private const val MAX_REPAIR_ATTEMPTS = 10

    private enum class Operation {
        UNION, INTERSECTION, DIFFERENCE
    }

    private enum class Type {
        OUTSIDE, INSIDE, SAME_EDGE, DIFF_EDGE
    }

    private fun operation(
        ra: Region2,
        rb: Region2,
        operation: Operation,
        aPredicate: (Type) -> Boolean,
        bPredicate: (Type) -> Boolean
    ): Region2 {
        val split: Split.Result = Split.split(ra, rb)
        val a: Region2 = split.a
        val b: Region2 = split.b

        // Partition rings into arcs separated at intersection points
        val pa: List<Arc> = partition(a, split.splits)
        var pb: List<Arc> = partition(b, split.splits)
        if (operation == Operation.DIFFERENCE) {
            pb = pb.map { obj: Arc -> obj.reverse() }
        }

        // Filter out arcs which are to be ignored, per our operation
        var arcs: MutableSet<Arc> = mutableSetOf()
        pa.filter { arc: Arc -> aPredicate(classify(b, arc)) }.forEach { value -> arcs.add(value) }
        pb.filter { arc: Arc -> bPredicate(classify(a, arc)) }.forEach { value -> arcs.add(value) }

        /*
    describe("split", split.splits.elements());
    describe("arcs", arcs.elements().stream().map(Arc::vertices).toArray(IList[]::new));
    VERTICES.forEach(v -> System.out.println(VERTICES.indexOf(v) + " " + v));
    // */
        val result = mutableListOf<Ring2>()
        val consumed = mutableSetOf<Arc>()

        // First we're going to extract complete cycles, and then try to iteratively repair the graph
        for (i in 0 until MAX_REPAIR_ATTEMPTS) {
            // Construct a graph where the edges are the set of all arcs connecting the vertices
            val graph = DirectedGraph<Vec2, Set<Arc>>()
            arcs.forEach { arc: Arc ->
                graph.link(
                    arc.head(), arc.tail(), mutableSetOf(arc)
                ) { obj, s -> obj.union(s) }
            }

            //graph.vertices().forEach(v -> System.out.println(VERTICES.indexOf(v) + " " + graph.out(v).stream().map(VERTICES::indexOf).collect(Lists.linearCollector())));
            if (i > 0) {
                for (path in repairGraph(
                    graph,
                    (pa + pb) - arcs - consumed
                )) {
                    for (arc in path) {
                        // if the graph currently contains the arc, remove it
                        if (arcs.contains(arc)) {
                            //describe("remove", arc.vertices());
                            graph.unlink(arc.head(), arc.tail())
                            arcs.remove(arc)

                            // if the graph doesn't contain the arc, add it
                        } else {
                            //describe("add", arc.vertices());
                            graph.link(arc.head(), arc.tail(), mutableSetOf(arc))
                            arcs.add(arc)
                        }
                    }
                }
            }

            // find every cycle in the graph, and then expand those cycles into every possible arc combination, yielding a bunch
            // of rings ordered from largest to smallest
            val cycles: List<List<Arc>> = Graphs.cycles(graph)
                .map { cycle ->
                    edges(cycle
                    ) { x, y -> graph.edge(x, y).toList() }
                }
                .map { paths -> Combinatorics.combinations(paths) }
                .flatten()
                .sortedBy { area -> area(area) }
                .reversed()

            // extract as many cycles as possible without using the same arc twice
            for (cycle in cycles) {
                //describe("cycle", cycle.stream().map(Arc::vertices).toArray(IList[]::new));
                if (cycle.any { value -> consumed.contains(value) }
                ) {
                    continue
                }
                cycle.forEach { value -> consumed.add(value) }
                result.add(ring(cycle))
            }
            arcs = (arcs - consumed).toMutableSet()
            if (arcs.size == 0) {
                break
            }
        }
        //assert(arcs.size() == 0L)
        return Region2(result)
    }

    private fun isTop(c: Curve2): Boolean {
//        if (c == null) {
//            return false
//        }
        val delta: Double = c.end().x - c.start().x
        return if (delta == 0.0) {
            c.end().y > c.start().y
        } else delta < 0
    }


    private fun classify(
        region: Region2,
        arc: Arc
    ): Type {
        // we want some point near the middle of the arc which is unlikely to coincide with a vertex, because those
        // sometimes sit ambiguously on the edge of the other region
        val result: Ring2.Result = region.test(arc.position(1.0 / E))
        return if (!result.inside) {
            Type.OUTSIDE
        } else if (result.curve == null) {
            Type.INSIDE
        } else {
            if (isTop(arc.first()) == isTop(result.curve))
                Type.SAME_EDGE else Type.DIFF_EDGE
        }
    }

    /**
     * Cuts the rings of a region at the specified vertices, yielding a list of arcs that will serve as the edges of our
     * graph.
     */
    private fun partition(
        region: Region2,
        vertices: Set<Vec2>
    ): List<Arc> {
        val result: MutableList<Arc> = mutableListOf()
        for (r in region.rings) {
            val cs: Array<Curve2> = r.curves
            var offset = 0
            while (offset < cs.size) {
                if (vertices.contains(cs[offset].start())) {
                    break
                }
                offset++
            }
            if (offset == cs.size) {
                result.add(Arc(cs.toMutableList()))
            } else {
                var acc = Arc()
                for (i in offset until cs.size) {
                    val c: Curve2 = cs[i]
                    if (vertices.contains(c.start())) {
                        if (acc.size > 0) {
                            result.add(acc)
                        }
                        acc = Arc(mutableListOf(c))
                    } else {
                        acc.add(c)
                    }
                }
                for (i in 0 until offset) {
                    acc.add(cs[i])
                }
                if (acc.size > 0) {
                    result.add(acc)
                }
            }
        }
        return result
    }

    private val SHORTEST_ARC = { x: Arc, y: Arc -> if (x.length() < y.length()) x else y }

    private fun repairGraph(
        graph: DirectedGraph<Vec2, Set<Arc>>,
        unused: Iterable<Arc>
    ): List<List<Arc>> {

        // create a graph of all the unused arcs
        val search = DirectedGraph<Vec2, Arc>()
        for (arc in unused) {
            search.link(arc.head(), arc.tail(), arc, SHORTEST_ARC)
        }

        // add in the existing arcs as reversed edges, so we can potentially retract them
        for (e in graph.edges()) {
            val arc: Arc = e.value().minByOrNull(
                { obj -> obj.length() })!!

            search.link(arc.tail(), arc.head(), arc, SHORTEST_ARC)
        }

        //search.vertices().forEach(v -> System.out.println(VERTICES.indexOf(v) + " " + search.out(v).stream().map(VERTICES::indexOf).collect(Lists.linearCollector())));
        //graph.vertices().forEach(v -> System.out.println(VERTICES.indexOf(v) + " " + graph.out(v).stream().map(VERTICES::indexOf).collect(Lists.linearCollector())));
        val `in` = graph.vertices()
            .filter { v -> graph.`in`(v).isEmpty() }.toSet()
        val out = graph.vertices()
            .filter { v -> graph.out(v).isEmpty() }.toSet()
        val currIn = (`in` + setOf()).toMutableSet()
        val currOut = (out + setOf()).toMutableSet()

        // attempt to greedily pair our outs and ins
        val result = mutableSetOf<List<Arc>>()
        while (currIn.size > 0 && currOut.size > 0) {
            val path = Graphs.shortestPath(search, currOut,
                { value -> `in`.contains(value) },
                { e -> e.value().length() }
            )
            // if our search found a vertex that was previously claimed, we need something better than a greedy search
            if (path == null || !currIn.contains(path.last())) {
                break
            } else {
                currOut.remove(path.first())
                currIn.remove(path.last())
                result.add(edges(path) { from, to -> search.edge(from, to) })
            }
        }
        return if (currIn.size == 0 || currOut.size == 0) {
            result.toList()
        } else Combinatorics.permutations(out.toList())
            .map { vs -> greedyPairing(search, vs, `in`) }
            .minByOrNull { path -> path.sumOf { arcs -> length(arcs) } }
            .orEmpty()

        // Do greedy pairings with every possible vertex ordering, and choose the one that results in the shortest aggregate
        // paths. If `out` is sufficiently large, `permutations` will just return a subset of random shufflings, and it's
        // possible we won't find a single workable solution this time around.
    }

    private fun greedyPairing(
        graph: DirectedGraph<Vec2, Arc>,
        out: List<Vec2>,
        `in`: Set<Vec2>
    ): List<List<Arc>> {
        val result: MutableList<List<Arc>> = mutableListOf()
        val currIn = (`in` + setOf()).toMutableSet()
        for (v in out) {
            // this will only happen if a vertex needs to have multiple edges added/removed, but we'll just get it on the
            // next time around
            if (currIn.size == 0) {
                break
            }
            val path: List<Vec2> = Graphs.shortestPath(graph, listOf(v),
                { value -> currIn.contains(value) },
                { e -> e.value().length() }) ?: (return emptyList())
            currIn.remove(path.last())
            result.add(edges(path) { from, to -> graph.edge(from, to) })
        }
        return result
    }

    private fun <U, V> edges(vertices: List<U>, edge: (U, U) -> V): List<V> {
        val result = mutableListOf<V>()
        for (i in 0 until vertices.size - 1) {
            result.add(edge(vertices[i], vertices[(i + 1)]))
        }
        return result
    }

    private fun area(arcs: List<Arc>): Double {
        return abs(arcs.sumOf { arc -> arc.signedArea() })
    }

    private fun length(arcs: List<Arc>): Double {
        return abs(arcs.sumOf { arc -> arc.length() })
    }

    private fun ring(arcs: List<Arc>): Ring2 {
        val acc: MutableList<Curve2> = mutableListOf()
        arcs.forEach { arc ->
            arc.forEach { value -> acc.add(value) }
        }
        return Ring2(acc)
    }

    ///
    fun union(a: Region2, b: Region2): Region2 {
        return operation(a, b,
            Operation.UNION,
            { t -> t == Type.OUTSIDE || t == Type.SAME_EDGE },
             { t: Type -> t == Type.OUTSIDE })
    }

    fun intersection(a: Region2, b: Region2): Region2 {
        return operation(a, b,
            Operation.INTERSECTION,
             { t: Type -> t == Type.INSIDE || t == Type.SAME_EDGE },
             { t: Type -> t == Type.INSIDE })
    }

    fun difference(a: Region2, b: Region2): Region2 {
        return operation(a, b,
            Operation.DIFFERENCE,
             { t: Type -> t == Type.OUTSIDE || t == Type.DIFF_EDGE },
             { t: Type -> t == Type.INSIDE })
    }
}